src/server/string_family.cc

```diff
@@ -787,9 +787,19 @@ void StringFamily::Set(CmdArgList args, ConnectionContext* cntx) {
       if (abs_ms < 0)
         return cntx->SendError(InvalidExpireTime("set"));
 
-      // Redis reports just OK in this case
-      if (rel_ms < 0)
+      // Remove existed key if the key is expired already
+      if (rel_ms < 0) {
+        cntx->transaction->ScheduleSingleHop([key](const Transaction* tx, EngineShard* es) {
```

Comment:
This is also a bug in redis `SET` (because as you noticed there could be the case that an object already exists without expiry and it won't be deleted). However, (I think) we might be able to skip this code and fall back `SetGeneric(cntx, sparams, key, value);` because it should be able to handle already expired values. If it doesn't then it might make sense to to handle this case there instead and not (explicitly) here?

Reply:
In redis if key is exist without expiration and you set with expired time the key will be deleted

Reply:
That's what I am saying. But instead of issuing a single hop that deletes the key here you can really in the operation below. That way you have a single path of execution that handles all the cases

Reply:
this code should not be here. Again, lets not confuse parsing and execution.
For redis SET command, there are other options like NX, XX that affect how the command is executed. For example, with NX and past time, I assume it should be a NOOP.  Therefore, I think you should move the logic into SetCmd code.

Reply:
Also, to reduce code duplication, I suggest that you expose OpDel in generic_family.h similarly to how we expose OpExists there.

Reply:
I can not move this logic into setcmd code because we generate a SET command and after that execute it as Redis command

Reply:
I am not sure I understand why it can not be moved to SetCmd code.
you generate `SET key val EXAT ...` for memcached flow. It will be executed inside SetCmd code. The same is true for the Redis SET command.

Reply:
I need to calculate is it relative or unit time before I create SET command and after that it will be executed as SetCmd

Reply:
@romange what if we call SetCmd with a timevalue that is already expired? My suggestion is to handle this operation in db_slice (I am not sure what happens if you call AddOrFind with a value that is already expired but it's worth checking). In other words my idea is, when you try to insert an expired value you just deleted. That way the flow above in SetCmd remains the same. 

Also I agree with @romange we should have one flow, and do not execute the Hop here (that's what I meant in my OG comment above)

Reply:
We can make a separate fix, for all SET family commands because there will be more changes and it's 2 different tasks

Reply:
Well it's already merged so it's a separate task anyway, I am just thinking out loud